### ISSUE TITLE
write/truncate in defineFile breaks when writing a value shorter than what is already cached.

### ISSUE THREAD

**Issue by buckley-w-david** [osnr/TabFS#40]
1. navigate to https://llanfairpwllgwyngyllgogerychwyrndrobwllllantysiliogogogoch.co.uk/
2. change to that tabs directory.
```
[david@darvid-pc TabFS]$ ls -lah mnt/tabs/by-title
total 0
drwxrwxrwx 3 root root  0 Dec 31  1969 .
drwxr-xr-x 3 root root  0 Dec 31  1969 ..
lr--r--r-- 1 root root 12 Dec 31  1969 Llanfairpwll_-_Llanfair_PG_in_Anglesey__North_Wales__Llanfairpwllgwyngyllgogerychwyrndrobwllllantysiliogogogoch_29 -> ../by-id/29
lr--r--r-- 1 root root 12 Dec 31  1969 New_Issue___osnr_TabFS_32 -> ../by-id/32
[david@darvid-pc TabFS]$ cd mnt/tabs/by-id/29/
[david@darvid-pc 29]$ cat url.txt 
https://llanfairpwllgwyngyllgogerychwyrndrobwllllantysiliogogogoch.co.uk/
```
3. write something short to url.txt
```
[david@darvid-pc 29]$ echo "http://example.com" > url.txt 
[david@darvid-pc 29]$ 
```
4. read url back
```
[david@darvid-pc 29]$ cat url.txt
http://example.comlgwyngyllgogerychwyrndrobwllllantysiliogogogoch.co.uk/
[david@darvid-pc 29]$ 
```
As you can see, instead of replacing the contents with "http://example.com", it has only overwritten the first `n` characters (where `n` is the length of the url that is written).

I'm not sure, but I believe this is because truncate (and write?) do not affect the value that was cached when the file was opened.

---

logs

cd into directory
```
req 
Object { id: 1228281408, op: "getattr", path: "/" }
background.js:646:11
resp 
Object { st_mode: 16877, st_nlink: 3, st_size: 0, op: "getattr" }
background.js:674:13
req 
Object { id: 1236813376, op: "getattr", path: "/tabs" }
background.js:646:11
resp 
Object { st_mode: 16877, st_nlink: 3, st_size: 0, op: "getattr" }
background.js:674:13
req 
Object { id: 1228281408, op: "getattr", path: "/tabs/by-id" }
background.js:646:11
resp 
Object { st_mode: 16877, st_nlink: 3, st_size: 0, op: "getattr" }
background.js:674:13
req 
Object { id: 1236813376, op: "getattr", path: "/tabs/by-id/29" }
background.js:646:11
resp 
Object { st_mode: 16877, st_nlink: 3, st_size: 0, op: "getattr" }
background.js:674:13
```
cat url.txt
```
req 
Object { id: 1228281408, op: "getattr", path: "/tabs/by-id/29/cat" }
background.js:646:11
UnixError: 
    UnixError moz-extension://a20c96b9-c645-476c-b5a7-4e0810d50246/background.js:25
    findRoute moz-extension://a20c96b9-c645-476c-b5a7-4e0810d50246/background.js:637
    onMessage moz-extension://a20c96b9-c645-476c-b5a7-4e0810d50246/background.js:658
background.js:663:13
resp 
Object { op: "getattr", error: 2 }
background.js:674:13
req 
Object { id: 1236813376, op: "getattr", path: "/tabs/by-id/29/url.txt" }
background.js:646:11
resp 
Object { st_mode: 33206, st_nlink: 1, st_size: 74, op: "getattr" }
background.js:674:13
req 
Object { id: 1228281408, op: "open", path: "/tabs/by-id/29/url.txt", flags: 32768 }
background.js:646:11
resp 
Object { fh: 1, op: "open" }
background.js:674:13
req 
Object { id: 1236813376, op: "read", path: "/tabs/by-id/29/url.txt", size: 131072, offset: 0, fh: 1, flags: 32768 }
background.js:646:11
resp 
Object { buf: "aHR0cHM6Ly9sbGFuZmFpcnB3bGxnd3luZ3lsbGdvZ2VyeWNod3lybmRyb2J3bGxsbGFudHlzaWxpb2dvZ29nb2NoLmNvLnVrLwo=", op: "read" }
background.js:674:13
req 
Object { id: 1228281408, op: "read", path: "/tabs/by-id/29/url.txt", size: 131072, offset: 74, fh: 1, flags: 32768 }
background.js:646:11
resp 
Object { buf: "", op: "read" }
background.js:674:13
req 
Object { id: 1228281408, op: "release", path: "/tabs/by-id/29/url.txt", fh: 1 }
background.js:646:11
resp 
Object { op: "release" }
background.js:674:13
```
echo "http://example.com" > url.txt
```
req 
Object { id: 1236813376, op: "getattr", path: "/tabs/by-id/29/url.txt" }
background.js:646:11
resp 
Object { st_mode: 33206, st_nlink: 1, st_size: 74, op: "getattr" }
background.js:674:13
req 
Object { id: 1228281408, op: "open", path: "/tabs/by-id/29/url.txt", flags: 32769 }
background.js:646:11
resp 
Object { fh: 2, op: "open" }
background.js:674:13
req 
Object { id: 1228281408, op: "truncate", path: "/tabs/by-id/29/url.txt", size: 0 }
background.js:646:11
resp 
Object { op: "truncate" }
background.js:674:13
req 
Object { id: 1228281408, op: "getattr", path: "/tabs/by-id/29/url.txt" }
background.js:646:11
resp 
Object { st_mode: 33206, st_nlink: 1, st_size: 74, op: "getattr" }
background.js:674:13
req 
Object { id: 1236813376, op: "write", path: "/tabs/by-id/29/url.txt", buf: "http://example.com\n", offset: 0, fh: 2, flags: 32769 }
background.js:646:11
resp 
Object { size: 19, op: "write" }
background.js:674:13
req 
Object { id: 1228281408, op: "release", path: "/tabs/by-id/29/url.txt", fh: 2 }
background.js:646:11
resp 
Object { op: "release" }
```
cat url.txt
```
resp 
Object { op: "getattr", error: 2 }
background.js:674:13
req 
Object { id: 1228281408, op: "getattr", path: "/tabs/by-id/29/url.txt" }
background.js:646:11
resp 
Object { st_mode: 33206, st_nlink: 1, st_size: 73, op: "getattr" }
background.js:674:13
req 
Object { id: 1236813376, op: "open", path: "/tabs/by-id/29/url.txt", flags: 32768 }
background.js:646:11
resp 
Object { fh: 3, op: "open" }
background.js:674:13
req 
Object { id: 1228281408, op: "read", path: "/tabs/by-id/29/url.txt", size: 131072, offset: 0, fh: 3, flags: 32768 }
background.js:646:11
resp 
Object { buf: "aHR0cDovL2V4YW1wbGUuY29tbGd3eW5neWxsZ29nZXJ5Y2h3eXJuZHJvYndsbGxsYW50eXNpbGlvZ29nb2dvY2guY28udWsvCg==", op: "read" }
background.js:674:13
req 
Object { id: 1236813376, op: "read", path: "/tabs/by-id/29/url.txt", size: 131072, offset: 73, fh: 3, flags: 32768 }
background.js:646:11
resp 
Object { buf: "", op: "read" }
background.js:674:13
req 
Object { id: 1228281408, op: "release", path: "/tabs/by-id/29/url.txt", fh: 3 }
background.js:646:11
resp 
Object { op: "release" }
```

**Comment by 4z3** [comment]
Looks like this issue is fixable with fuse3, which [passes `struct fuse_file_info *fi` to `fuse_fs_truncate`](https://github.com/libfuse/libfuse/blob/fuse-3.10.1/lib/fuse.c#L2141-L2142), enabling us to use the file handle to truncate open (cached) objects.

**Comment by 4z3** [comment]
FWIW, a patch that uses fuse3 but has been _only tested on Linux_ can be found [here](https://github.com/4z3/TabFS/commit/a36e95f).

**Comment by osnr** [comment]
Cool. I need to look into whether this works on macFUSE and get back to you. On Linux, does it buy us anything beyond #52? Or could we roll back #52 given your patch?

**Comment by 4z3** [comment]
> On Linux, does it buy us anything beyond #52? Or could we roll back #52 given your patch?

Using fuse3 is unrelated to #52.  What it does is providing the possibility to take care of the TODO in `defineFile`'s `truncate` function, e.g. like this _(using `Cache.setObjectForHandle` introduced by #52)_:

```diff
--- a/extension/background.js
+++ b/extension/background.js
@@ -170,17 +170,17 @@ const defineFile = (getData, setData) => ({
   },
   async release({fh}) { Cache.removeObjectForHandle(fh); return {}; },
 
-  async truncate({path, size}) {
-    // TODO: weird case if they truncate while the file is open
-    // (but `echo hi > foo.txt`, the main thing I care about, uses
-    // O_TRUNC which thankfully doesn't do that)
-    let arr = toUtf8Array(await getData(path));
-    if (size !== arr.length) {
-      const newArr = new Uint8Array(size);
-      newArr.set(arr.slice(0, Math.min(size, arr.length)));
-      arr = newArr;
+  async truncate({path, fh, size}) {
+    let arr;
+    if (typeof fh === "number") {
+      arr = Cache.getObjectForHandle(fh);
+    } else {
+      arr = toUtf8Array(await getData(path));
     }
-    await setData(path, utf8ArrayToString(arr)); return {};
+    const newArr = new Uint8Array(size);
+    newArr.set(arr.slice(0, Math.min(arr.length, size)));
+    Cache.setObjectForHandle(fh, newArr);
+    await setData(path, utf8ArrayToString(newArr)); return {};
   }
 });
```

With this patch applied to the `truncate` function, it's possible to use tools that truncate an open file.  E.g. it would solve the issue of `echo`ing a shorter URL to `url.txt` [initially described](https://github.com/osnr/TabFS/issues/40#issue-779783134) (without relying on `FUSE_CAP_ATOMIC_O_TRUNC`), and it would allow using tools that always truncate an open file, like e.g. Vim does.

In other words: let's keep #52, and I can open PRs for both using fuse3 and the patched truncate function, but I'd only do that if fuse3 is available for Mac.
